libs/grok/src/main/java/org/elasticsearch/grok/ThreadInterrupter.java

```diff
@@ -0,0 +1,134 @@
+     * @param scheduler             A scheduler that is able to execute a command for each fixed interval
+     */
+    static ThreadInterrupter newInstance(long interval,
+                                         long maxExecutionTime,
+                                         LongSupplier relativeTimeSupplier,
+                                         BiFunction<Long, Runnable, ScheduledFuture<?>> scheduler) {
+        return new Default(interval, maxExecutionTime, relativeTimeSupplier, scheduler);
+    }
+    
+    /**
+     * @return A noop implementation that does not interrupt threads and is useful for testing and pre-defined grok expressions.
+     */
+    static ThreadInterrupter noop() {
+        return new Noop();
+    }
+    
+    class Noop implements ThreadInterrupter {
+        
+        private Noop() {
+        }
+        
+        @Override
+        public void register() {
+        }
+        
+        @Override
+        public void deregister() {
+        }
+    }
+    
+    class Default implements ThreadInterrupter {
+        
+        private final long interval;
+        private final long maxExecutionTime;
+        private final LongSupplier relativeTimeSupplier;
+        private final BiFunction<Long, Runnable, ScheduledFuture<?>> scheduler;
+        final ConcurrentHashMap<Thread, Long> registry = new ConcurrentHashMap<>();
+        
+        private Default(long interval,
+                        long maxExecutionTime,
+                        LongSupplier relativeTimeSupplier,
+                        BiFunction<Long, Runnable, ScheduledFuture<?>> scheduler) {
+            this.interval = interval;
+            this.maxExecutionTime = maxExecutionTime;
+            this.relativeTimeSupplier = relativeTimeSupplier;
+            this.scheduler = scheduler;
+            scheduler.apply(interval, this::interruptLongRunningExecutions);
+        }
+        
+        public void register() {
+            Long previousValue = registry.put(Thread.currentThread(), relativeTimeSupplier.getAsLong());
+            assert previousValue == null;
+        }
+        
+        public void deregister() {
+            Long previousValue = registry.remove(Thread.currentThread());
+            assert previousValue != null;
+        }
+        
+        private void interruptLongRunningExecutions() {
+            try {
+                final long currentRelativeTime = relativeTimeSupplier.getAsLong();
+                for (Map.Entry<Thread, Long> entry : registry.entrySet()) {
+                    long threadTime = entry.getValue();
+                    if ((currentRelativeTime - threadTime) > maxExecutionTime) {
+                        entry.getKey().interrupt();
+                        // not removing the entry here, this happens in the deregister() method.
+                    }
+                }
+            } finally {
+                scheduler.apply(interval, this::interruptLongRunningExecutions);
```

Comment:
This is potentially risky. What if the `try` block throws an `OutOfMemoryError` and the `scheduler#apply` throws a `RejectedExecutionException`? We lose the `OutOfMemoryError` and miss dying with dignity. Now, the `scheduler#apply` should really on throw a `RejectedExecutionException` if the scheduler is shutdown which should only happen when we are shutting down, but let's not take any chances here that we lose a fatal error?